### ISSUE TITLE
More user friendly args than dclass/sclass

### ISSUE THREAD

**Issue by thehunmonkgroup** [yassen-itlabs/py-linux-traffic-control#11]
Wondering if there is a more user friendly way to express the source/destination port stuff?

It seems pretty easy to get confused on when to use sclass vs dclass in either the ingress or egress shaping.

I'd even advocate for more user friendly terms for egress/ingress -- eg., from the point of view of the server doing the shaping, I think upload/download would be easier to comprehend.

**Comment by yassen-itlabs** [comment]
How about something like:

``# ltc tc --clear --dev eth0 --down --dport tcp:8000-8099:768kbit:5% --sport udp:all:384kbit``

Thus using ``--down`` (or a longer ``--download`` if that's too short) for ingress, ``--up`` (or ``--upload``) would be for egress, and maybe we should require it (now the default is egress).

Another option would be to put the dport/sport qualifier into the class specs, like this:

``# ltc tc --clear --dev eth0 --down --disc tcp:dport:8000-8099:768kbit:5% --disc udp:sport:all:384kbit``

Note that I propose also changing the``--iface`` option into ``--dev`` as it is more in-line with the iproute2 terminology.

@thehunmonkgroup : Chad, please let me know your thoughts, thanks.

**Comment by thehunmonkgroup** [comment]
i actually have some fairly specific preferences on this, based on what i've experienced as 'standard' command line argument syntax:

- args should provide two forms, both a single letter (-c) and a full text (--clear) version
  - -d / --download
  - -u / --upload
  - -i / --interface
 - i propose that we eliminate the switch for the current protocol:port:rate string, eg. ```ltc tc --clear --interface eth0 --download tcp:dport:8000-8099:768kbit:5% udp:sport:all:384kbit```
 - i personally like --interface instead of --dev, especially since it eliminates the issue of two args that should have a -d single letter. however, i'm not attached to that position
 - i do think --upload / --download should both be required, and mutually exclusive, i believe this makes it easier to validate the syntax, and harder for users to make usage mistakes

**Comment by yassen-itlabs** [comment]
The idea to have the class spec directives (like ``tcp:dport:8000-8099:768kbit:5%``) as multiple options to ``--download`` (or ``--upload`` resp.) seems great! I'll try to implement this as soon as I get the refactoring done (or at least the essential part of it).

Q: Why have ``--download`` and ``--upload`` being mutually exclusive? I think they can happily coexist in a single command, if the user wants/needs to combine them. We need to enforce the restriction that at least one of these is there. We will have the nice side effect that ``pyltc tc --clear -u -d`` will clear both ingress and egress chains. If we start to support ``--show``, we will have the same nice side effect there as well -- ``pyltc tc --show -u -d`` will show status of both chains. What do you think?

**Comment by thehunmonkgroup** [comment]
how will the arg  parser work in that case -- will it accept multiple space separated protocol:location:port:rate args as belonging to the last passed switch until/unless another switch is passed?

**Comment by yassen-itlabs** [comment]
yep, everything coming after a ``--download`` switch will be interpreted as ingress-related; everything after ``--upload`` as egress-related.  The profile config file will look like this:

```
[3g-sym]
clear
iface eth0  # the primary interface
download
  tcp:sport:8000-8088:96kbit
  udp:sport:5000-5080:96kbit:3%
  udp:dport:20000-39999:256kbit:1%
upload
  tcp:dport:8080:128kbit
```
The indent is purely for better readability.

**Comment by thehunmonkgroup** [comment]
yep, looks good, just change iface to interface for that long arg

**Comment by yassen-itlabs** [comment]
Absolutely. This one has slipped by accident ;)

**Comment by yassen-itlabs** [comment]
I believe this is fully resolved with 90c2e9669eb6ebbc8fe173fd0cf360a176358f41. We now support arguments exactly as specified here; in addition, we support simultaneous handling of ``--upload`` and ``--download`` options. (Also, the ``all`` port range keyword is preserved.) Closing.
